File: other-programming-languages/PHP/2024/2024-008/2024-008.php
<?php

/**
 * Reads the grid map from a file.
 * @param string $filename The name of the file containing the grid map.
 * @return array A list of strings representing the grid.
 */
function readMap($filename) {
    return array_map('rtrim', file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES));
}

/**
 * Computes unique antinode positions using the pairwise method.
 * @param array $grid A list of strings representing the grid map.
 * @return array A set of unique antinode positions as strings.
 */
function computeAntinodesPairwise($grid) {
    $rows = count($grid);
    $cols = $rows > 0 ? strlen($grid[0]) : 0;

    // Group antennas by frequency
    $antennasByFreq = [];
    for ($r = 0; $r < $rows; $r++) {
        for ($c = 0; $c < $cols; $c++) {
            $ch = $grid[$r][$c];
            if ($ch !== '.') {
                if (!isset($antennasByFreq[$ch])) {
                    $antennasByFreq[$ch] = [];
                }
                $antennasByFreq[$ch][] = [$r, $c];
            }
        }
    }

    $antinodes = [];

    foreach ($antennasByFreq as $freq => $coords) {
        $n = count($coords);
        if ($n < 2) continue;

        for ($i = 0; $i < $n; $i++) {
            list($rA, $cA) = $coords[$i];
            for ($j = $i + 1; $j < $n; $j++) {
                list($rB, $cB) = $coords[$j];

                // Compute P1 = 2B - A
                $p1_r = 2 * $rB - $rA;
                $p1_c = 2 * $cB - $cA;
                if ($p1_r >= 0 && $p1_r < $rows && $p1_c >= 0 && $p1_c < $cols) {
                    $antinodes["$p1_r,$p1_c"] = true;
                }

                // Compute P2 = 2A - B
                $p2_r = 2 * $rA - $rB;
                $p2_c = 2 * $cA - $cB;
                if ($p2_r >= 0 && $p2_r < $rows && $p2_c >= 0 && $p2_c < $cols) {
                    $antinodes["$p2_r,$p2_c"] = true;
                }
            }
        }
    }
    return array_keys($antinodes);
}

/**
 * Computes unique antinode positions using the line-drawing method.
 * @param array $grid A list of strings representing the grid map.
 * @return array A set of unique antinode positions as strings.
 */
function computeAntinodesLines($grid) {
    $rows = count($grid);
    $cols = $rows > 0 ? strlen($grid[0]) : 0;

    // Group antennas by frequency
    $antennasByFreq = [];
    for ($r = 0; $r < $rows; $r++) {
        for ($c = 0; $c < $cols; $c++) {
            $ch = $grid[$r][$c];
            if ($ch !== '.') {
                if (!isset($antennasByFreq[$ch])) {
                    $antennasByFreq[$ch] = [];
                }
                $antennasByFreq[$ch][] = [$r, $c];
            }
        }
    }

    $antinodes = [];

    foreach ($antennasByFreq as $freq => $coords) {
        $n = count($coords);
        if ($n < 2) continue;

        for ($i = 0; $i < $n; $i++) {
            list($rA, $cA) = $coords[$i];
            for ($j = $i + 1; $j < $n; $j++) {
                list($rB, $cB) = $coords[$j];
                addLinePoints($rA, $cA, $rB, $cB, $rows, $cols, $antinodes);
            }
        }
    }
    return array_keys($antinodes);
}

/**
 * Adds all points along a line between two antennas to the antinodes set.
 * @param int $rA Row of the first antenna.
 * @param int $cA Column of the first antenna.
 * @param int $rB Row of the second antenna.
 * @param int $cB Column of the second antenna.
 * @param int $rows Number of rows in the grid.
 * @param int $cols Number of columns in the grid.
 * @param array $antinodes Array to store unique antinode positions.
 */
function addLinePoints($rA, $cA, $rB, $cB, $rows, $cols, &$antinodes) {
    $dr = $rB - $rA;
    $dc = $cB - $cA;
    $g = gcd(abs($dr), abs($dc));
    $dr /= $g;
    $dc /= $g;

    // Add points in the forward direction
    $rP = $rA;
    $cP = $cA;
    while ($rP >= 0 && $rP < $rows && $cP >= 0 && $cP < $cols) {
        $antinodes["$rP,$cP"] = true;
        $rP += $dr;
        $cP += $dc;
    }

    // Add points in the backward direction
    $rP = $rA - $dr;
    $cP = $cA - $dc;
    while ($rP >= 0 && $rP < $rows && $cP >= 0 && $cP < $cols) {
        $antinodes["$rP,$cP"] = true;
        $rP -= $dr;
        $cP -= $dc;
    }
}

/**
 * Computes the greatest common divisor of two integers.
 * @param int $a First integer.
 * @param int $b Second integer.
 * @return int The greatest common divisor of a and b.
 */
function gcd($a, $b) {
    return $b === 0 ? $a : gcd($b, $a % $b);
}

/**
 * Formats the elapsed time in seconds with nanoseconds as a human-readable string.
 * @param float $seconds The elapsed time in seconds.
 * @return string The formatted time as a string.
 */
function formatTime($seconds) {
    return number_format($seconds, 9) . ' s';
}

/**
 * Main function to run the program.
 */
function main() {
    $filename = 'input.txt';
    $grid = readMap($filename);

    $startTime = microtime(true);

    // Part 1: Compute using Pairwise method
    $pairwiseStart = microtime(true);
    $pairwiseAntinodes = computeAntinodesPairwise($grid);
    $pairwiseEnd = microtime(true);
    echo "Pairwise: " . formatTime($pairwiseEnd - $pairwiseStart) . "\n";
    echo "Number of unique antinodes (Pairwise method): " . count($pairwiseAntinodes) . "\n";

    // Part 2: Compute using Line-drawing method
    $lineStart = microtime(true);
    $lineAntinodes = computeAntinodesLines($grid);
    $lineEnd = microtime(true);
    echo "Line-drawing: " . formatTime($lineEnd - $lineStart) . "\n";
    echo "Number of unique antinodes (Line-drawing method): " . count($lineAntinodes) . "\n";

    $endTime = microtime(true);
    echo "Overall: " . formatTime($endTime - $startTime) . "\n";
}

main();

File: other-programming-languages/Ruby/2024/2024-008/2024-008.rb
require 'set'

# Main class for computing antinodes using two methods: pairwise and line-drawing.
class AntinodeCalculator
  # Reads the grid map from a file.
  #
  # @param filename [String] The name of the file containing the grid map.
  # @return [Array<String>] An array of strings representing the grid.
  def self.read_map(filename)
    File.readlines(filename, chomp: true)
  end

  # Computes unique antinode positions using the pairwise method.
  #
  # @param grid [Array<String>] An array of strings representing the grid map.
  # @return [Set<Array<Integer>>] A set of unique antinode positions.
  def self.compute_antinodes_pairwise(grid)
    rows = grid.size
    cols = rows.positive? ? grid[0].size : 0

    # Group antennas by frequency
    antennas_by_freq = Hash.new { |h, k| h[k] = [] }
    grid.each_with_index do |row, r|
      row.each_char.with_index do |ch, c|
        antennas_by_freq[ch] << [r, c] unless ch == '.'
      end
    end

    antinodes = Set.new

    antennas_by_freq.each_value do |coords|
      n = coords.size
      next if n < 2

      coords.each_with_index do |a, i|
        r_a, c_a = a
        (i + 1...n).each do |j|
          b = coords[j]
          r_b, c_b = b

          # Compute P1 = 2B - A
          p1_r, p1_c = 2 * r_b - r_a, 2 * c_b - c_a
          antinodes.add([p1_r, p1_c]) if p1_r.between?(0, rows - 1) && p1_c.between?(0, cols - 1)

          # Compute P2 = 2A - B
          p2_r, p2_c = 2 * r_a - r_b, 2 * c_a - c_b
          antinodes.add([p2_r, p2_c]) if p2_r.between?(0, rows - 1) && p2_c.between?(0, cols - 1)
        end
      end
    end

    antinodes
  end

  # Computes unique antinode positions using the line-drawing method.
  #
  # @param grid [Array<String>] An array of strings representing the grid map.
  # @return [Set<Array<Integer>>] A set of unique antinode positions.
  def self.compute_antinodes_lines(grid)
    rows = grid.size
    cols = rows.positive? ? grid[0].size : 0

    # Group antennas by frequency
    antennas_by_freq = Hash.new { |h, k| h[k] = [] }
    grid.each_with_index do |row, r|
      row.each_char.with_index do |ch, c|
        antennas_by_freq[ch] << [r, c] unless ch == '.'
      end
    end

    antinodes = Set.new

    antennas_by_freq.each_value do |coords|
      n = coords.size
      next if n < 2

      coords.each_with_index do |a, i|
        r_a, c_a = a
        (i + 1...n).each do |j|
          r_b, c_b = coords[j]
          add_line_points(r_a, c_a, r_b, c_b, rows, cols, antinodes)
        end
      end
    end

    antinodes
  end

  # Adds all points along a line between two antennas to the antinodes set.
  #
  # @param r_a [Integer] Row of the first antenna.
  # @param c_a [Integer] Column of the first antenna.
  # @param r_b [Integer] Row of the second antenna.
  # @param c_b [Integer] Column of the second antenna.
  # @param rows [Integer] Number of rows in the grid.
  # @param cols [Integer] Number of columns in the grid.
  # @param antinodes [Set<Array<Integer>>] Set to store unique antinode positions.
  def self.add_line_points(r_a, c_a, r_b, c_b, rows, cols, antinodes)
    dr = r_b - r_a
    dc = c_b - c_a
    g = gcd(dr.abs, dc.abs)
    dr /= g
    dc /= g

    # Add points in the forward direction
    r_p, c_p = r_a, c_a
    while r_p.between?(0, rows - 1) && c_p.between?(0, cols - 1)
      antinodes.add([r_p, c_p])
      r_p += dr
      c_p += dc
    end

    # Add points in the backward direction
    r_p, c_p = r_a - dr, c_a - dc
    while r_p.between?(0, rows - 1) && c_p.between?(0, cols - 1)
      antinodes.add([r_p, c_p])
      r_p -= dr
      c_p -= dc
    end
  end

  # Computes the greatest common divisor of two integers.
  #
  # @param a [Integer] First integer.
  # @param b [Integer] Second integer.
  # @return [Integer] The greatest common divisor of a and b.
  def self.gcd(a, b)
    b.zero? ? a : gcd(b, a % b)
  end

  # Main function to run the program.
  #
  # @param filename [String] Input file name containing the grid.
  def self.main(filename)
    grid = read_map(filename)

    overall_start = Time.now

    # Part 1: Compute using Pairwise method
    start_time = Time.now
    pairwise_antinodes = compute_antinodes_pairwise(grid)
    part1_time = Time.now - start_time
    puts "Part 001 finished in #{format_time(part1_time)}"
    puts "Number of unique antinodes (Pairwise method): #{pairwise_antinodes.size}"

    # Part 2: Compute using Line-drawing method
    start_time = Time.now
    line_antinodes = compute_antinodes_lines(grid)
    part2_time = Time.now - start_time
    puts "Part 002 finished in #{format_time(part2_time)}"
    puts "Number of unique antinodes (Line-drawing method): #{line_antinodes.size}"

    # Compute total computation time
    computation_total_time = part1_time + part2_time
    puts "Total computation time: #{format_time(computation_total_time)}"

    # Compute overall runtime (includes overhead)
    overall_total_time = Time.now - overall_start
    puts "Overall total time: #{format_time(overall_total_time)}"
  end

  # Formats the elapsed time as a human-readable string.
  #
  # @param seconds [Float] The elapsed time in seconds.
  # @return [String] The formatted time as a string.
  def self.format_time(seconds)
    format('%.9f s', seconds)
  end
end

# Run the program with the input file
AntinodeCalculator.main('\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-008\\input.txt')
